crates/turborepo-auth/src/lib.rs

```diff
@@ -12,19 +12,81 @@ mod ui;
 pub use auth::*;
 pub use error::Error;
 pub use login_server::*;
+use turborepo_api_client::TokenClient;
+use turborepo_vercel_api::token::ResponseTokenMetadata;
 
 /// Token is the result of a successful login. It contains the token string and
-/// a boolean indicating whether the token already existed on the filesystem.
-#[derive(Debug)]
+/// potentially metadata about the token.
+#[derive(Debug, Clone)]
 pub enum Token {
+    /// An existing token on the filesystem
     Existing(String),
+    /// A token that was just created, but not yet written to the filesystem
     New(String),
 }
 impl Token {
-    pub fn into_inner(self) -> String {
+    pub fn new(token: String) -> Self {
+        Self::New(token)
+    }
+    pub fn existing(token: String) -> Self {
+        Self::Existing(token)
+    }
+    /// Checks if the token is valid. We do a few checks:
+    /// 1. Fetch the token metadata.
+    /// 2. From the metadata, check if the token is active.
+    /// 3. If the token is a SAML SSO token, check if it's expired.
+    pub async fn is_valid(&self, client: &impl TokenClient) -> Result<bool, Error> {
```

Comment:
This is a great candidate for a unit test to verify the logic and make sure:
 - we catch case where all scopes are expired
 - we catch case where `active_at` is after the current time
 - any token with a saml_sso scope is considered to be invalid

Reply:
Refactored this to put the heavy lifting in `is_token_active`, since in theory an expired token isn't active.